interop-cats/shared/src/main/scala/zio/stream/cats.scala

```diff
@@ -0,0 +1,122 @@
+package zio.stream.interop
+
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import zio.interop.test.CatsTestFunctions
+
+import cats._
+import cats.arrow._
+import zio._
+import zio.stream._
+
+object catz extends CatsInstances
+
+abstract class CatsInstances extends CatsInstances1 {
+  implicit def alternativeInstance[R, E]: Alternative[ZStream[R, E, *]] =
+    new CatsAlternative[R, E]
+}
+
+abstract class CatsInstances1 extends CatsInstances2 {
+  implicit def monoidKInstance[R, E]: MonoidK[ZStream[R, E, *]] =
+    new CatsMonoidK[R, E]
+
+  implicit def bifunctorInstance[R]: Bifunctor[ZStream[R, *, *]] =
+    new CatsBifunctor[R] {}
+
+  implicit def zioArrowInstance[E]: ArrowChoice[ZStream[*, E, *]] = new CatsArrow[E]
+}
+
+sealed abstract class CatsInstances2 extends CatsInstances3 {
+  implicit def semigroupKInstance[R, E: Semigroup]: SemigroupK[ZStream[R, E, *]] =
+    new CatsSemigroupK[R, E]
+}
+
+sealed abstract class CatsInstances3 {
+  implicit def monadErrorInstance[R, E]: MonadError[ZStream[R, E, *], E] =
+    new CatsMonadError[R, E]
+}
+
+private class CatsAlternative[R, E] extends CatsMonoidK[R, E] with CatsApplicative[R, E]
+
+private class CatsMonadError[R, E] extends CatsMonad[R, E] with MonadError[ZStream[R, E, *], E] {
+  override final def handleErrorWith[A](fa: ZStream[R, E, A])(f: E => ZStream[R, E, A]): ZStream[R, E, A] =
+    fa.catchAll(f)
+  override final def raiseError[A](e: E): ZStream[R, E, A] = ZStream.fail(e)
+
+  override def attempt[A](fa: ZStream[R, E, A]): ZIO[R, E, Either[E, A]] = fa.either
+}
+
+private class CatsMonad[R, E]
+    extends CatsApplicative[R, E]
+    with Monad[ZStream[R, E, *]]
+    with StackSafeMonad[ZStream[R, E, *]] {
+  override final def flatMap[A, B](fa: ZStream[R, E, A])(f: A => ZStream[R, E, B]): ZStream[R, E, B] = fa.flatMap(f)
+
+  override final def widen[A, B >: A](fa: ZStream[R, E, A]): ZStream[R, E, B] = fa
+  override final def map2[A, B, Z](fa: ZStream[R, E, A], fb: ZStream[R, E, B])(f: (A, B) => Z): ZStream[R, E, Z] =
+    fa.zipWith(fb) { case (Some(a), Some(b)) => Some(f(a, b)); case _ => None }
```

Comment:
Pretty sure there's a law that says `map2` (and `ap`) need to be consistent with `flatMap`. This function applies the function pointwise on the streams, but to be consistent with `flatMap` we need the cartesian product. That's `crossWith`.

Reply:
Got it 👍